Fix #2942 Sets "Depicts" (P180 property) automatically for images uploaded via "nearby"

VitalyVPinchuk opened the pull request:
Sets "Depicts" (P180 property) automatically for images uploaded via "nearby".

Fixes #2942

Added two calls to Commons API for getting `pageid` for filename (`getFileEntityId()`) and setting P180 property for that `pageid` (`wikidataEditEntity()`)

**Tests performed (required)**
Tested ProdDebug on emulator with API level 28.

nicolas-raoul commented on app/src/main/java/fr/free/nrw/commons/mwapi/ApacheHttpClientMediaWikiApi.java:
```diff
@@ -436,6 +447,88 @@ public String wikidataCreateClaim(String entityId, String property, String snakt
+     */
+    @Nullable
+    @Override
+    public String wikidataEditEntity(String entityId, String fileEntityId) throws IOException {
```
Why does the name of this method include `wikidata`? The P180 property must be sent to Commons' Wikibase server, not to the Wikidata server, right? Would not a name like `commonsWikibaseEditEntity` or `wikibaseEditEntity` make more sense? If it is for consistency with the existing methods, feel free to rename them from `wikidata` to `wikibase` _(up to now we did not connect to any Wikibase server other than the Wikidata server, which explains the current wording)_. Thanks :-)

VitalyVPinchuk replied:
Sure.

ashishkumar468 commented on app/src/main/java/fr/free/nrw/commons/mwapi/ApacheHttpClientMediaWikiApi.java:
```diff
@@ -436,6 +447,88 @@ public String wikidataCreateClaim(String entityId, String property, String snakt
+    @Nullable
+    @Override
+    public String wikidataEditEntity(String entityId, String fileEntityId) throws IOException {
+
```
Please remove the code comment lines, if not needed anymore :-)

VitalyVPinchuk replied:
Yes, sure.

vanshikaarora commented:
Thanks for the changes @VitalyVPinchuk, I have merged your PR with my branch and will test it:)

VitalyVPinchuk commented:
> New APIs should not be added in `ApacheHttpClientMediaWikiApi`.
> 
> Either use APIs from the [data-client](https://github.com/wikimedia/wikimedia-android-data-client) library or check the [library's documentation](https://github.com/wikimedia/wikimedia-android-data-client) to see how custom APIs should be added.
> 
> @VitalyVPinchuk
> 
> cc @vanshikaarora
> 
> For reference check #2998 or #2894

Hello @maskaravivek !
Unfortunately, I'm not aware neither of new API nor of old one.
Thanks for the links, I'll try to figure it out.

maskaravivek commented:
> Hello @maskaravivek !
> Unfortunately, I'm not aware neither of new API nor of old one.
> Thanks for the links, I'll try to figure it out.

From new APIs I simply mean any API that we add in our codebase.

nicolas-raoul commented:
@VitalyVPinchuk Your PR is generating a lot of interest :-)

Sorry for not mentioning that earlier, but as Vivek said we should use the new (more modern and maintained) library https://github.com/wikimedia/wikimedia-android-data-client

Many parts of the app already use it, so you can see how it works. It is extensible via custom API calls if needed.

Thanks a lot! :-)

VitalyVPinchuk commented:
Hello.
I've tried to switch to wikimedia-android-data-client but can't get `editToken` to make `wbeditentity` call. That token was retrieved by `public String getEditToken()` method.
Would anyone help me? I pushed the changes.

vanshikaarora commented:
@VitalyVPinchuk in my code that uses the same library I have added :

```
Observable.fromCallable(new Callable<String>() {
            @Override public String call() throws Exception {
                return mediaWikiApi.getEditToken();
            }
        }).subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread()).subscribe(new Consumer<String>() {
            @Override public void accept(String s) throws Exception {
               //creating the map
             // you may call your function function here using the retrofit interface that you have created
                Observable.fromCallable(() -> captionInterface.addLabelstoWikidata(map))
                        .subscribeOn(Schedulers.io())
                        .observeOn(AndroidSchedulers.mainThread())
                        .subscribe(revisionId ->{

                                    revisionId.enqueue(new Callback<MwQueryResponse>() {
                                        @Override
                                        public void onResponse(Call<MwQueryResponse> call, Response<MwQueryResponse> response) {
                                            MwQueryResponse result = response.body();
                                            Timber.e((""+response.isSuccessful()));
                                        }

                                        @Override
                                        public void onFailure(Call<MwQueryResponse> call, Throwable t) {

                                            call.cancel();
                                        }
                                    });
                                },

                                throwable -> {
                                    //error message
                                });
            }
                });
```


Hope this helps :)

vanshikaarora commented:
Hey @VitalyVPinchuk thanks for the code :).

I was also using your code to upload depictions in my branch. Can you please tell how can I upload multiple depictions using these functions

VitalyVPinchuk commented:
Hi @vanshikaarora 
You shouldn't blindly rely on that code cause I'm new to android sdk, just forewarning you :)

I think you can set multiple tags for an item (file or other entity) in a single call. Just list the tags in "claims" array of "data" parameter.
But I don't know how to set a tag for multiple items (entities) in a single call.

I found these help pages which may be useful:
https://www.wikidata.org/w/api.php?action=help&modules=wbeditentity
and the related sandbox
https://www.wikidata.org/wiki/Special:ApiSandbox#action=wbeditentity&id=Q42&data=%7B%22claims%22:%5B%7B%22mainsnak%22:%7B%22snaktype%22:%22value%22,%22property%22:%22P56%22,%22datavalue%22:%7B%22value%22:%22ExampleString%22,%22type%22:%22string%22%7D%7D,%22type%22:%22statement%22,%22rank%22:%22normal%22%7D%5D%7D

vanshikaarora commented:
>I think you can set multiple tags for an item (file or other entity) in a single call. Just list the tags in "claims" array of "data" parameter.

Thanks @VitalyVPinchuk I'll look through. Also in your `addPropertyP180` function in which Json object do you pass the tag that is to be uploaded

VitalyVPinchuk commented:
Hello!
I still can't get centralauthtoken on beta cluster with an error  
> badtoken The centralauthtoken is not valid.  
But it was working on the prod-cluster where I've been banned for multiple uploads :) 

Would anyone please look into the code to get things going?

vanshikaarora commented:
Hey @VitalyVPinchuk thanks for the code :-)
Were you able to upload depictions using dummy value of `centralauthtoken`?

VitalyVPinchuk commented:
> Hey @VitalyVPinchuk thanks for the code :-)
> Were you able to upload depictions using dummy value of `centralauthtoken`?

Hi @vanshikaarora 
Yes, I think dummy value works on beta cluster. Using "+\\" for posting depictions doesn't complain about wrong `centralauthtoken` but gives another error, which says that there's no such `entityId`, which I presume could probably be true.

vanshikaarora commented:
@VitalyVPinchuk I see that this branch is getting 
>E/WikidataEditService: Error occurred while setting DEPICTS property
    org.wikipedia.dataclient.mwapi.MwException: The "token" parameter must be set.

Can you please post here what token are you passing in this line https://github.com/commons-app/apps-android-commons/blob/9059a1af10e88c5a6400b2c0d921fb5533e6175e/app/src/main/java/fr/free/nrw/commons/upload/WikiBaseInterface.java#L24

Also can you please post your request and complete response here.

Thanks :-)

VitalyVPinchuk commented:
@vanshikaarora 
The token is dummy "+\\" as I test it on beta flavor (real value doesn't work for me here).
fileEntityId is, for example, M66123 
and data is 
{"claims":[{"mainsnak":{"snaktype":"value","property":"P245962","datavalue":{"value":{"entity-type":"item","numeric-id":"10","id":"Q10"},"type":"wikibase-entityid"}},"type":"statement","rank":"preferred"}]}

Response throws exception.

vanshikaarora commented:
The function here is annotated with `@FormUrlEndoding`

https://github.com/commons-app/apps-android-commons/blob/9059a1af10e88c5a6400b2c0d921fb5533e6175e/app/src/main/java/fr/free/nrw/commons/upload/WikiBaseInterface.java#L23

May be using encoded `+/` might work.
Also can you please post the API request from your logcat here? :-)

VitalyVPinchuk commented:
I think this is it

Request

```
2019-08-30 17:14:18.936 7354-7619/fr.free.nrw.commons.beta D/OkHttp: --> POST https://wikidata.org/w/api.php?format=json&formatversion=2&errorformat=plaintext&action=wbeditentity
2019-08-30 17:14:18.936 7354-7619/fr.free.nrw.commons.beta D/OkHttp: Content-Type: application/x-www-form-urlencoded
2019-08-30 17:14:18.938 7354-7619/fr.free.nrw.commons.beta D/OkHttp: Content-Length: 372
2019-08-30 17:14:18.938 7354-7619/fr.free.nrw.commons.beta D/OkHttp: Cache-Control: no-cache
2019-08-30 17:14:18.939 7354-7619/fr.free.nrw.commons.beta D/OkHttp: id=M0&token=%2B%5C&data=%7B%22claims%22%3A%5B%7B%22mainsnak%22%3A%7B%22snaktype%22%3A%22value%22%2C%22property%22%3A%22P245962%22%2C%22datavalue%22%3A%7B%22value%22%3A%7B%22entity-type%22%3A%22item%22%2C%22numeric-id%22%3A%2210%22%2C%22id%22%3A%22Q10%22%7D%2C%22type%22%3A%22wikibase-entityid%22%7D%7D%2C%22type%22%3A%22statement%22%2C%22rank%22%3A%22preferred%22%7D%5D%7D
2019-08-30 17:14:18.939 7354-7619/fr.free.nrw.commons.beta D/OkHttp: --> END POST (372-byte body)
```
Response
```
2019-08-30 17:14:19.292 7354-7619/fr.free.nrw.commons.beta D/OkHttp: <-- 200 https://www.wikidata.org/w/api.php?format=json&formatversion=2&errorformat=plaintext&action=wbeditentity (353ms)
2019-08-30 17:14:19.292 7354-7619/fr.free.nrw.commons.beta D/OkHttp: date: Fri, 30 Aug 2019 14:14:20 GMT
2019-08-30 17:14:19.292 7354-7619/fr.free.nrw.commons.beta D/OkHttp: content-type: application/json; charset=utf-8
2019-08-30 17:14:19.292 7354-7619/fr.free.nrw.commons.beta D/OkHttp: server: mw1279.eqiad.wmnet
2019-08-30 17:14:19.292 7354-7619/fr.free.nrw.commons.beta D/OkHttp: x-powered-by: HHVM/3.18.6-dev
2019-08-30 17:14:19.292 7354-7619/fr.free.nrw.commons.beta D/OkHttp: mediawiki-api-error: notoken
2019-08-30 17:14:19.292 7354-7619/fr.free.nrw.commons.beta D/OkHttp: p3p: CP="This is not a P3P policy! See https://www.wikidata.org/wiki/Special:CentralAutoLogin/P3P for more info."
2019-08-30 17:14:19.292 7354-7619/fr.free.nrw.commons.beta D/OkHttp: cache-control: private, must-revalidate, max-age=0
2019-08-30 17:14:19.292 7354-7619/fr.free.nrw.commons.beta D/OkHttp: content-disposition: inline; filename=api-result.json
2019-08-30 17:14:19.292 7354-7619/fr.free.nrw.commons.beta D/OkHttp: x-content-type-options: nosniff
2019-08-30 17:14:19.292 7354-7619/fr.free.nrw.commons.beta D/OkHttp: x-frame-options: DENY
2019-08-30 17:14:19.292 7354-7619/fr.free.nrw.commons.beta D/OkHttp: backend-timing: D=58765 t=1567174460494006
2019-08-30 17:14:19.293 7354-7619/fr.free.nrw.commons.beta D/OkHttp: vary: Accept-Encoding,X-Seven
2019-08-30 17:14:19.293 7354-7619/fr.free.nrw.commons.beta D/OkHttp: x-varnish: 84336580, 684982671, 1058101686
2019-08-30 17:14:19.293 7354-7619/fr.free.nrw.commons.beta D/OkHttp: accept-ranges: bytes
2019-08-30 17:14:19.293 7354-7619/fr.free.nrw.commons.beta D/OkHttp: age: 0
2019-08-30 17:14:19.293 7354-7619/fr.free.nrw.commons.beta D/OkHttp: x-cache: cp1089 pass, cp3030 pass, cp3043 pass
2019-08-30 17:14:19.293 7354-7619/fr.free.nrw.commons.beta D/OkHttp: x-cache-status: pass
2019-08-30 17:14:19.293 7354-7619/fr.free.nrw.commons.beta D/OkHttp: server-timing: cache;desc="pass"
2019-08-30 17:14:19.293 7354-7619/fr.free.nrw.commons.beta D/OkHttp: strict-transport-security: max-age=106384710; includeSubDomains; preload
2019-08-30 17:14:19.293 7354-7619/fr.free.nrw.commons.beta D/OkHttp: x-analytics: WMF-Last-Access=30-Aug-2019;WMF-Last-Access-Global=30-Aug-2019;https=1
2019-08-30 17:14:19.293 7354-7619/fr.free.nrw.commons.beta D/OkHttp: x-client-ip: 188.65.238.121
2019-08-30 17:14:19.295 7354-7619/fr.free.nrw.commons.beta D/OkHttp: {"errors":[{"code":"notoken","text":"The \"token\" parameter must be set.","module":"wbeditentity"}],"docref":"See https://www.wikidata.org/w/api.php for API usage. Subscribe to the mediawiki-api-announce mailing list at &lt;https://lists.wikimedia.org/mailman/listinfo/mediawiki-api-announce&gt; for notice of API deprecations and breaking changes.","servedby":"mw1279"}
2019-08-30 17:14:19.295 7354-7619/fr.free.nrw.commons.beta D/OkHttp: <-- END HTTP (371-byte body)
```

vanshikaarora commented:
@VitalyVPinchuk I see that your API does not contains the token. Can you please check :-)
```
 https://wikidata.org/w/api.php?format=json&formatversion=2&errorformat=plaintext&action=wbeditentity
```

Also I would suggest that instead of uploading an image and trying the network response.
You should first verify that the URL :  https://wikidata.org/w/api.php?format=json&formatversion=2&errorformat=plaintext&action=wbeditentity
gives you the desired response

VitalyVPinchuk commented:
But this is POST not GET.
The data comes in the body of the message.
Or am I wrong?

vanshikaarora commented:
> But this is POST not GET.
> The data comes in the body of the message.
> Or am I wrong?

Yes you are correct but at least the token format error can be figured out this way

nicolas-raoul commented:
> But this is POST not GET.

You can send POST requests with the curl command. It is convenient because it is fast, reproducible, and produces only the necessary output :-)

VitalyVPinchuk commented:
Thanks. I'll try.

VitalyVPinchuk commented:
Hello @vanshikaarora 
It looks as if the problem was because of dependency injection performed upon wikidata not commons wiki site at NetworkingModule.java. So I changed it to
```
    @Provides
    @Singleton
    public WikiBaseInterface provideWikiBaseInterface(@Named(NAMED_COMMONS_WIKI_SITE) WikiSite commonsWikiSite) {
        return ServiceFactory.get(commonsWikiSite, BuildConfig.COMMONS_URL, WikiBaseInterface.class);
    }
```

And when I updated from 0.0.18 to 0.0.25 version of `com.dmitrybrant:wikimedia-android-data-client` which is used in your case, it required the following line of code prior to calling`wikiBaseClient.postEditEntity(...)` in order to make `csrfTokenClient.getTokenBlocking()` work:
        `Observable.defer((Callable<ObservableSource<Boolean>>) () ->`

What does it do and how could it be connected with getting `editToken`? I have no idea.

VitalyVPinchuk commented:
Hi!
Is there anything else I should do about this PR?

vanshikaarora commented:
Hey @VitalyVPinchuk . Thanks for your work:-)

I am currently having my mid-semester Examination. I will test the PR in the coming week. Sorry for the delay.

And thanks again for you contribution :-)

vanshikaarora commented:
@VitalyVPinchuk was there any need to change the version from  0.0.18 to 0.0.25.
Also with these changes are you able to retrieve correct `editToken`?

VitalyVPinchuk commented:
> @VitalyVPinchuk was there any need to change the version from 0.0.18 to 0.0.25.
> Also with these changes are you able to retrieve correct `editToken`?

Hello @vanshikaarora 
As far as I remember I didn't manage to retrieve correct `editToken` for beta cluster so I had to overwrite it with `+/` when using 0.0.18. But it was correct with 0.0.25, so I decided to use this version.

vanshikaarora commented:
Okay, is the PR ready for testing, Or you still have some issues to address? :-)

VitalyVPinchuk commented:
It would be nice if someone could test it

vanshikaarora commented on app/src/main/java/fr/free/nrw/commons/di/NetworkingModule.java:
```diff
@@ -113,14 +136,14 @@ public Gson provideGson() {
     @Provides
     @Singleton
-    public ReviewInterface provideReviewInterface(@Named("commons-wikisite") WikiSite commonsWikiSite) {
-        return ServiceFactory.get(commonsWikiSite, BuildConfig.COMMONS_URL, ReviewInterface.class);
```
oh, okay

macgills commented:
These changes are on the structured data branch
